### ISSUE TITLE
Using SFT trainer and collator: 'eval_loss': nan

### ISSUE THREAD

**Issue by Sosycs** [huggingface/trl#1053]
Hello, 

Thank you very much for your great work.
I am not sure what could make this happen, I got the following output:
`{'eval_loss': nan, 'eval_runtime': 143.8323, 'eval_samples_per_second': 17.576, 'eval_steps_per_second': 2.197}`

those are my training arguments and settings:
```
local_rank = -1
per_device_train_batch_size = 4
per_device_eval_batch_size = 4
gradient_accumulation_steps = 1
learning_rate = 2e-4 #9.65e-6
max_grad_norm = 0.3
weight_decay = 0.001
lora_alpha = 16
lora_dropout = 0.1
lora_r = 64
max_seq_length = 512
use_4bit = True
fp16 = False
bf16 = True
packing = False
gradient_checkpointing = True
optim = "paged_adamw_32bit"
lr_scheduler_type = "cosine"
max_steps = -1
warmup_ratio = 0.03
group_by_length = True
save_steps = 10
logging_steps = 1

training_arguments = TrainingArguments(
    output_dir=output_dir,
    per_device_train_batch_size=per_device_train_batch_size,
    gradient_accumulation_steps=gradient_accumulation_steps,
    optim=optim,
    save_steps=save_steps,
    logging_steps=logging_steps,
    learning_rate=learning_rate,
    fp16=fp16,
    bf16=bf16,
    max_grad_norm=max_grad_norm,
    max_steps=max_steps,
    warmup_ratio=warmup_ratio,
    group_by_length=group_by_length,
    lr_scheduler_type=lr_scheduler_type,
    report_to="tensorboard",
    auto_find_batch_size=True,
)


trainer = SFTTrainer(
    model=finetuned_model,
    eval_dataset=val_dataset,
    peft_config=peft_config,
    max_seq_length=max_seq_length,
    tokenizer=tokenizer,
    args=training_arguments,
    packing=packing,
    data_collator=DataCollatorForCompletionOnlyLM(response_template="[/INST]" , tokenizer=tokenizer),
    formatting_func=formatting_func,
)

```

Is this related to the use of collator?

**Comment by lvwerra** [comment]
Maybe @younesbelkada has an idea.

**Comment by ethan-young-vgm** [comment]
I am having a similar issue except the eval_loss is not even present:
{'eval_runtime': 248.0972, 'eval_samples_per_second': 1.266, 'eval_steps_per_second': 1.266, 'epoch': 0.03}

This happens whether or not I train on completions only. The training loss shows up fine though when logging.
Training args:
    train_on_source: False,
    adam8bit: False
    double_quant: True,
    quant_type: nf4",
    bits: 4
    lora_r: 64
    lora_alpha: 16
    lora_dropout: 0.1
    report_to: tensorboard
    output_dir: './output'
    optim: 'paged_adamw_32bit'
    per_device_train_batch_size: 1
    per_device_eval_batch_size : 8 
    gradient_accumulation_steps: 1
    max_steps: -1 
    weight_decay: 0.01
    learning_rate: 1e-5
    remove_unused_columns: False
    max_grad_norm: 0.3
    gradient_checkpointing: True
    lr_scheduler_type: constant
    warmup_ratio: 0.03
    logging_steps: 20
    group_by_length: True
    save_strategy: 'no'
    evaluation_strategy: 'steps'
    eval_steps: 20
    num_train_epochs: 4
    neftune_noise_alpha: 5.0

**Comment by Morizeyao** [comment]
same problem here

**Comment by younesbelkada** [comment]
Hmm I am not sure what is wrong here :/ do you face the same issue with 0.7.8? Also what happens if you don't use that collator?

**Comment by Morizeyao** [comment]
I found out that nan is caused by DataCollatorForCompletionOnlyLM not finding response template position.
Changing template solved the problem for me.

**Comment by younesbelkada** [comment]
@Morizeyao how did you changed the template ?

**Comment by Morizeyao** [comment]
Although the template text did existed in the training sample text, after tokenize, it's not in tokenized training sample. That caused DataCollatorForCompletionOnlyLM can't find location.
I changed template to make it possible for code in collator to find location, so that solved the problem.

Before change: "Assistant: " (with space)
After change: "Assistant" (only a word)

**Comment by lindsey98** [comment]
Hi, but if "Could not find response key `{self.response_template}", by right this instance is ignored in loss calculation?

https://github.com/huggingface/trl/blob/main/trl/trainer/utils.py#L129
```
 warnings.warn(
                        f"Could not find response key `{self.response_template}` in the "
                        f'following instance: {self.tokenizer.decode(batch["input_ids"][i])} '
                        f"This instance will be ignored in loss calculation. "
                        f"Note, if this happens often, consider increasing the `max_seq_length`."
                    )
```

**Comment by hamzajameel** [comment]
So what happened, my max seq length was 512, and when ever the truncation was happening on those examples which had more than 512 token, the response template was also being truncated as well, so, basically it was truncating my label from prompt only for those example which had more than 512 tokens, now If increase the max seq length to 1024 the issue is Fixed !!! that's why the loss was nan for those example only cuz the label was truncated by setting the max seq == 512.

**Comment by lindsey98** [comment]
I write my own evaluator to skip the long examples where the response_template is being truncated, and this works:

```python

@torch.inference_mode()
def metric_eval(eval_dataloader, model):
    model.eval()
    nlls = []

    for i, batch in tqdm(enumerate(eval_dataloader)):
        if (batch["labels"] == -100).all():
            continue  # Skip this batch
        batch = {k: v.cuda() for k, v in batch.items()}
        with torch.amp.autocast("cuda", dtype=torch.bfloat16):
            out = model(**batch)
            neg_log_likelihood = out.loss

        nlls.append(neg_log_likelihood)

    # Log results at the end
    ppl = torch.exp(torch.stack(nlls).mean())
    print(ppl)

    return ppl
    
eval_loader = trainer.get_eval_dataloader()
metric_eval(eval_loader, model)

```

**Comment by ManyaWadhwa** [comment]
where did you make this code change? im facing the same issue where some instances from my val set are getting truncated cause being longer than the max seq length! thanks!

> I write my own evaluator to skip the long examples where the response_template is being truncated, and this works:
> 
> ```python
> @torch.inference_mode()
> def metric_eval(eval_dataloader, model):
>     model.eval()
>     nlls = []
> 
>     for i, batch in tqdm(enumerate(eval_dataloader)):
>         if (batch["labels"] == -100).all():
>             continue  # Skip this batch
>         batch = {k: v.cuda() for k, v in batch.items()}
>         with torch.amp.autocast("cuda", dtype=torch.bfloat16):
>             out = model(**batch)
>             neg_log_likelihood = out.loss
> 
>         nlls.append(neg_log_likelihood)
> 
>     # Log results at the end
>     ppl = torch.exp(torch.stack(nlls).mean())
>     print(ppl)
> 
>     return ppl
>     
> eval_loader = trainer.get_eval_dataloader()
> metric_eval(eval_loader, model)
> ```

**Comment by lindsey98** [comment]
Hi Manya, 
 I wrote this as an ad hoc fix during evaluation. I suppose we might need to overwrite compute_loss() in the Trainer class for a more systematic fix? For example:
  
  ```
  class CustomSFTTrainer(SFTTrainer):
     def compute_loss(self, model, inputs, return_outputs=False):
        ## TODO: rewrite the loss function with additional check on labels
     
        ......
        
  # Then create an instance of the trainer
  trainer = CustomSFTTrainer(
    model=finetuned_model,
    eval_dataset=val_dataset,
    peft_config=peft_config,
    max_seq_length=max_seq_length,
    tokenizer=tokenizer,
    args=training_arguments,
    packing=packing,
    data_collator=DataCollatorForCompletionOnlyLM(response_template="[/INST]" , tokenizer=tokenizer),
    formatting_func=formatting_func,
)
  ```

**Comment by joelvarunvasanth** [comment]
Hi all,

I get training loss to be a finite number while eval loss is Nan:

I experienced the same error with evaluation loss being Nan, and thanks to this issue, was able to figure out that the response_template was being truncated. However, if the problem is the truncation of the response_template, why then do I get the training loss to be a finite number (i.e., not Nan) all throughout the training?

**Comment by irfanmekic** [comment]
im also experiencing the same issue, using the DataCollatorForLanguageModeling with MLM=true - anyone have a fix?

**Comment by AiteYx** [comment]
> im also experiencing the same issue, using the DataCollatorForLanguageModeling with MLM=true - anyone have a fix?

The same message. The reason is the same as mentioned above. The data length exceeds the max-length of the model.

I filtered out the extremely long data.
```
def filter_short_examples(example):
    return len(tokenizer.apply_chat_template(example["prompt"] + example["completion"])) < training_args.max_length
filtered_dataset = dataset.filter(filter_short_examples)
```
